src/fstext/kaldi-fst-io.h

```diff
@@ -131,6 +161,86 @@ class VectorFstTplHolder {
 // Now make the original VectorFstHolder as the typedef o VectorFstHolder<StdArc>.
 typedef VectorFstTplHolder<StdArc> VectorFstHolder;
 
+// A generic Fst Holder which is similar with VectorFstHolder. Depending on the
+// FstType(), it will take different actions. This version currently supports 
+// ConstFst<StdArc> or VectorFst<StdArc>
+// This is a Holder class with T = Fst<Arc>, that meets the requirements
+// of a Holder class as described in ../util/kaldi-holder.h. This enables us to
+// read/write collections of FSTs indexed by strings, using the Table concept (
+// see ../util/kaldi-table.h).
+template<class Arc>
+class FstTplHolder {
+ public:
+  typedef Fst<Arc> T;
+
+  FstTplHolder(): t_(NULL) { }
+
+  static bool Write(std::ostream &os, bool binary, const T &t);
+
+  // Depending on the fst type, we cope t into this holder.
+  void Copy(const T &t) {
+    Clear();
+    if(t.Type() == "const") {
+      t_ = new ConstFst<Arc>(t);
+    } else if( t.Type() == "vector") {
+      t_ = new VectorFst<Arc>(t);
+    } else {
+      KALDI_WARN << "The version currently only support VectorFst and ConstFst.";
+      t_ = &t;
+    }
+  }
+
+  // Reads into the holder.
+  bool Read(std::istream &is);
+
+  // It's potentially a binary format, so must read in binary mode (linefeed
+  // translation will corrupt the file.  We don't know till we open the file if
+  // it's really binary, so we need to read in binary mode to be on the safe
+  // side.  Extra linefeeds won't matter, the text-mode reading code ignores
+  // them.
+  static bool IsReadInBinary() { return true; }
+
+  const T &Value() {
+    // code error if !t_.
+    if (!t_) KALDI_ERR << "FstTplHolder::Value() called wrongly.";
+    if(t_->Type() == "const") {
+      return *(dynamic_cast<ConstFst<Arc>*>(t_));
```

Comment:
There is no need to cast here- you are pointlessly casting and then implicitly casting back again.
And there is no need to check the type here.